elastic/elastic-agent
# Support log level setting from policy

## What does this PR do?

Support setting elastic-agent log level from fleet policy.
The agent will apply (in decreasing order of priority):
1. log level set specifically to the agent via `settings` action, if any
1. log level specified in fleet policy, if any
1. default hard-coded log level for elastic-agent

Whenever a `policy_change` or `settings` action is received, the settings action handler will reevaluate the loglevels specified and set the log level according to the priority above.



## Why is it important?

It allows users to manage elastic-agent verbosity easily through the fleet policy, while at the same time allowing to set a different log level to specific agents for troubleshooting issues.



## Checklist




## Author's Checklist



## How to test this PR locally



Create a simple policy in fleet and enroll an elastic-agent with that policy.
The start setting log levels in policy and for the specific agent using dev tools and the requests below

Set policy log level
```http
PUT kbn:/api/fleet/agent_policies/<policy id>
{
   "name": "<policy name>",
   "namespace": "default",
   "overrides": {
       "agent":{
         "logging": {
           "level": "error"
         }
       }
   }
}
```
Set log level for a specific agent
```http
POST kbn:/api/fleet/agents/<elastic agent id>/actions
{
  "action": {
    "type": "SETTINGS",
    "data": {
     "log_level": "debug"   
    }
  }
}
```

A good way to check the effect of the changes of log levels is to keep a terminal with `elastic-agent logs -f` command running: that way we can see the changes in elastic-agent logging in real-time.

Another way to check the current log level currently in use by agent is to use the `inspect` subcommand (grepping or using `yq` is recommended as the output is very verbose, for example:

`sudo elastic-agent inspect | yq .agent`

```yaml
download:
  sourceURI: https://artifacts.elastic.co/downloads/
features: null
headers: null
id: 25bd1b94-9b76-4a2d-bdd8-09d778e8cf44
logging:
  level: warning
monitoring:
  enabled: true
  http:
    buffer: null
    enabled: false
    host: localhost
    port: 6791
  logs: true
  metrics: true
  namespace: default
  use_output: default
protection:
  enabled: false
  signing_key: <redacted>
```

## Related issues


- Closes #2851

## Use cases



## Screenshots



## Logs



## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
